### PR TITLE
docs: Better documentation for developers

### LINKED ISSUE #375: Better documentation for developers
* Improve [guidelines](https://stdlib.safeds.com/en/stable/development/guidelines/)
  * Tests should be parametrized and should have IDs
  * Don't mutate objects but return new ones. Also explain how the docstring should be formatted
  * Sort `__all__` in `__init__.py` alphabetically
* Add other resources that would help onboarding new developers
  * Example: A checklist for code review (?)

### PR DESCRIPTION
**Pull request by zzril**
Closes #375.

### Summary of Changes

* Improved developer documentation about tests
* Added guidelines about copying objects (rather than modifying them in-place)
* Added code-style guidelines
* Added code review guidelines

### CODE REVIEW

File: docs/development/guidelines.md [hunk 1]
```diff
@@ -56,6 +56,43 @@ Some flag parameters drastically alter the semantics of a function. This can lea
     table.drop("name", axis="columns")
     ```
 
+### Return copies of objects
+
+Modifying objects in-place
+can lead to surprising behaviour
```
Comment (jxnior01) [inline docs/development/guidelines.md:62]:
```suggestion
can lead to surprising behaviours
```

Comment (zzril) [reply docs/development/guidelines.md:62]: I have always thought of "behaviour" as an uncountable noun. I would also, for instance, say that "a program shows unexpected behaviour" (not "*an* unexpected behaviour").
Are you sure about this?

**Review by jxnior01 — DISMISSED** [review]
Nice work!!

**Comment by zzril** [comment]
The `code style` section got somewhat small. Turns out we don't really have a common code style except for "do what the linter says".
Thought about alsl adding an FAQ for developers, but I came up with only like 3 questions, that didn't see emough, so I dropped the idea.

Don't have any other ideas, so let me mark this as ready for review.

File: docs/development/guidelines.md [hunk 2]
```diff
@@ -258,6 +274,49 @@ Passing values that are commonly used together around separately is tedious, ver
...
+when adding new classes to it.
+
+!!! success "**DO** (library code):"
+
+    ```py
+    __all__ = [
+    "Column",
+    "Row",
+    "Table",
+    "TaggedTable",
```
Comment (lars-reimann) [inline docs/development/guidelines.md:305]:
```suggestion
        "Column",
        "Row",
        "Table",
        "TaggedTable",
```

File: docs/development/guidelines.md [hunk 3]
```diff
@@ -258,6 +274,49 @@ Passing values that are commonly used together around separately is tedious, ver
...
+    ```
+
+!!! failure "**DON'T** (library code):"
+
+    ```py
+    __all__ = [
+    "Table",
+    "TaggedTable",
+    "Column",
+    "Row",
```
Comment (lars-reimann) [inline docs/development/guidelines.md:316]:
```suggestion
        "Table",
        "TaggedTable",
        "Column",
        "Row",
```

File: docs/development/guidelines.md [hunk 4]
```diff
@@ -370,7 +429,43 @@ Examples
 
 ## Tests
 
-If a function contains more code than just the getting or setting of a value, automated test should be added to the [`tests`][tests-folder] folder. The file structure in the tests folder should mirror the file structure of the [`src`][src-folder] folder.
+If a function contains more code
+than just the getting or setting of a value,
+automated test should be added
+to the [`tests`][tests-folder] folder.
```
Comment (lars-reimann) [inline docs/development/guidelines.md:435]: A little outdated now. Maybe instead write something like

> We aim for 100% line coverage, so for any new function automated tests ...

File: docs/development/guidelines.md [hunk 5]
```diff
@@ -370,7 +429,43 @@ Examples
...
+shall start with `test_should_`
+followed by a description
+of the expected behaviour,
+e.g. `test_should_add_column`.
+
+### Testcases
+
+Tests should be parametrized
+using `@pytest.mark.parametrize`,
+even if there is only a single testcase.
```
Comment (lars-reimann) [inline docs/development/guidelines.md:450]: Add reason: Easier to add new test cases in the future.

File: docs/development/guidelines.md [hunk 6]
```diff
@@ -370,7 +429,43 @@ Examples
...
+should mirror the file structure
+of the [`src`][src-folder] folder.
+
+Names of test functions
+shall start with `test_should_`
+followed by a description
+of the expected behaviour,
+e.g. `test_should_add_column`.
+
+### Testcases
```
Comment (lars-reimann) [inline docs/development/guidelines.md:446]:
```suggestion
### Parametrization
```

File: docs/development/guidelines.md (hunk 5 above)
Comment (lars-reimann) [inline docs/development/guidelines.md:450]:
```suggestion
even if there is only a single test case.
```

File: docs/development/guidelines.md [hunk 7]
```diff
@@ -370,7 +429,43 @@ Examples
...
+
+Tests should be parametrized
+using `@pytest.mark.parametrize`,
+even if there is only a single testcase.
+Testcases should be given
```
Comment (lars-reimann) [inline docs/development/guidelines.md:451]:
```suggestion
Test cases should be given
```

File: docs/development/guidelines.md [hunk 8]
```diff
@@ -370,7 +429,43 @@ Examples
 
 ## Tests
```
Comment (lars-reimann) [inline docs/development/guidelines.md:430]: It might be worth splitting this into some more subsections, including:

* File structure (with do/don't example?)
* Naming (with do/don't example?)
* Parametrization (exists already)

File: docs/development/guidelines.md [hunk 9]
```diff
@@ -370,7 +429,43 @@ Examples
...
+        [
+            ("table.csv", Table({"A": [1], "B": [2]})),
+            (Path("table.csv"), Table({"A": [1], "B": [2]})),
+            ("empty_table.csv", Table()),
+        ],
+        ids=["by string", "by path", "empty"],
+    )
+    def test_should_create_table_from_csv_file(path: str | Path, expected: Table) -> None:
+        ...
+    ```
```
Comment (lars-reimann) [inline docs/development/guidelines.md:468]: Maybe have an example specifically for the parametrization and using the do/don't schema from above?

File: docs/development/guidelines.md [hunk 10]
```diff
@@ -258,6 +274,49 @@ Passing values that are commonly used together around separately is tedious, ver
     training_feature_vectors, validation_feature_vectors, training_target_values, validation_target_values = split(feature_vectors, target_values)
     ```
 
+## Code style
```
Comment (lars-reimann) [inline docs/development/guidelines.md:277]: Better move this to the end of the document:

1. Design
2. Docstrings
3. Tests
4. Code style

Code style is the least important concern for our users, so it's also not as important for developers to keep in mind as the other items.

File: docs/development/code_review.md
Comment (lars-reimann) [inline docs/development/code_review.md:1]: You can also link to this document from the `Contributing` section of the `README`.

File: mkdocs.yml [hunk 11]
```diff
@@ -17,6 +17,7 @@ nav:
     - Environment: development/environment.md
     - Guidelines: development/guidelines.md
```
Comment (lars-reimann) [inline mkdocs.yml:18]: Should we call this `Project Guidelines` instead?

File: docs/development/code_review.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
+# Code review
+
+This document describes
+general guidelines for developers
+when reviewing a Pull Request (PR).
```
Comment (lars-reimann) [inline docs/development/code_review.md:5]: Afterwards you switch between writing `Pull Request` or `PR`. Use one or the other consistently.

File: docs/development/code_review.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+However, you should always:
+
+* Make sure all tests actually ran through (pytest, linter, code coverage)
+* Make sure that the branch is up-to-date with the `main` branch, so you actually test the behaviour that will result once the feature branch is merged
+
+## Check the Pull Request format
+
+* Check that the PR title starts with a fitting [type](https://github.com/Safe-DS/.github/blob/main/.github/CONTRIBUTING.md#types) (e.g. `feat`, `fix`, ...)
+* Check that the changes introduced by the PR are documented in the PR message
+* Check that the `time spent by team` has been updated on the issue page
```
Comment (lars-reimann) [inline docs/development/code_review.md:42]: Can be omitted, since it was relevant only for the PG.

File: docs/development/code_review.md (hunk 12 above)
Comment (lars-reimann) [inline docs/development/code_review.md:5]:
```suggestion
when reviewing a pull request (PR).
```

File: docs/development/code_review.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+### Understand the issue
+
+* Read the issue that was addressed by the PR and make sure you understand it
+* Read any discussions that occurred on the issue page
+
+### Check the Pull Request
+
+* Check the Pull Request for __completeness__: Was every point from the issue covered?
+* Check for __correctness__: Were the problems described in the issue solved in the intended way?
+* (For PRs that introduce new features:) Check the design - does it comply with our [guidelines][guidelines-general]?
```
Comment (lars-reimann) [inline docs/development/code_review.md:18]:
```suggestion
* (For PRs that introduce new features:) Check the design - does it comply with our [project guidelines][guidelines-general]?
```

In case we change the title of the document.

File: docs/development/code_review.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+* Check for potential bugs - edge cases, exceptions that may be raised in functions that were called etc.
+* Check the code style: Is it readable? Do variables have sensible names? Is it consistent to existing code? Is there a better / more elegant way to do certain things (e.g. f-string instead of manual string concatenation)?
```
Comment (lars-reimann) [inline docs/development/code_review.md:20]: Maybe link to the Python docs for the `f-string`.

File: docs/development/code_review.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+* Check any warnings reported by your IDE
+
+## Check the tests
+
+* Run the tests locally
+* Check if there are any warnings that are not caught by the test suite
+* Make sure the test cases are complete - do they cover all edge cases (e.g. empty tables)?
+* Make sure they comply to our [guidelines][guidelines-tests] - are the tests parametrized and do all testcases have descriptive IDs?
```
Comment (lars-reimann) [inline docs/development/code_review.md:28]:
```suggestion
* Make sure they comply to our [project guidelines][guidelines-tests] - are the tests parametrized and do all testcases have descriptive IDs?
```

In case we change the title of the document.

File: docs/development/code_review.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+This is largely covered by the automated tests.
+However, you should always:
+
+* Make sure all tests actually ran through (pytest, linter, code coverage)
+* Make sure that the branch is up-to-date with the `main` branch, so you actually test the behaviour that will result once the feature branch is merged
+
+## Check the Pull Request format
+
+* Check that the PR title starts with a fitting [type](https://github.com/Safe-DS/.github/blob/main/.github/CONTRIBUTING.md#types) (e.g. `feat`, `fix`, ...)
```
Comment (lars-reimann) [inline docs/development/code_review.md:40]:
```suggestion
* Check that the PR title starts with a fitting [type](https://github.com/Safe-DS/.github/blob/main/.github/CONTRIBUTING.md#types) (e.g. `feat`, `fix`, `docs`, ...)
```

Just because `docs` is also needed often.

File: docs/development/code_review.md [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+select the `Approve` option
+to approve the changes made by the PR.
+If you like, you can add a "LGTM" meme
+from [lgtm.party](https://lgtm.party/),
+preferably one with a cat image.
+
+If no other reviewer suggested changes either,
+please also mark the corresponding issue
+as `Ready to merge` on the issue's page
+to update its status on the issue board.
```
Comment (lars-reimann) [inline docs/development/code_review.md:71]: Can be omitted since it was just needed for the PG.

File: docs/development/code_review.md [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+If you found any problems,
+select `Request Changes` instead.
+In this case, please also
+mark the corresponding issue
+as `In progress` again.
```
Comment (lars-reimann) [inline docs/development/code_review.md:77]: Can be omitted since it was just neeeded for the PG.

**Review by lars-reimann — CHANGES_REQUESTED** [review]
Looks great overall, just some minor suggestions.

File: docs/development/code_review.md (hunk 19 above)
Comment (zzril) [reply docs/development/code_review.md]: Removed it.

File: docs/development/code_review.md (hunk 18 above)
Comment (zzril) [reply docs/development/code_review.md]: Removed.

File: docs/development/code_review.md (hunk 15 above)
Comment (zzril) [reply docs/development/code_review.md]: Added a link.

File: docs/development/guidelines.md (hunk 5 above)
Comment (zzril) [reply docs/development/guidelines.md]: Done.

File: docs/development/guidelines.md (hunk 4 above)
Comment (zzril) [reply docs/development/guidelines.md]: Changed it.

File: docs/development/code_review.md
Comment (zzril) [reply docs/development/code_review.md:1]: Added a link.

Comment (zzril) [reply docs/development/code_review.md:1]: On that occasion: When hosting the docs locally and following the link from the README there, I get a strange-looking, oversized (possibly optimized for mobile?) page without CSS.
Do we actually have a proper 404 page?

File: mkdocs.yml (hunk 11 above)
Comment (zzril) [reply mkdocs.yml]: I think it makes sense. Changed it.

File: docs/development/guidelines.md (hunk 8 above)
Comment (zzril) [reply docs/development/guidelines.md:387]: Splitted it into the suggested subsections and added a do / don't example for naming.

For the file structure, I'm not sure how such an example should look like. We cannot show Python code as in all the other examples.
At the same time, it should be fairly clear to understand, so I'm not sure examples are necessary here?

File: docs/development/code_review.md
Comment (lars-reimann) [reply docs/development/code_review.md:1]: We do but unfortunately Readthedocs won't show it properly (see #176). We can't do much about that for now.

File: docs/development/code_review.md [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+to approve the changes made by the PR.
+If you like, you can add a "LGTM" meme
+from [lgtm.party](https://lgtm.party/),
+preferably one with a cat image.
+
+If you found any problems,
+select `Request Changes` instead.
+
+[guidelines-general]: https://stdlib.safeds.com/en/stable/development/guidelines/
+[guidelines-tests]: https://stdlib.safeds.com/en/stable/development/guidelines/#tests
```
Comment (lars-reimann) [inline docs/development/code_review.md:71]: Should use relative links:

```suggestion
[guidelines-general]: guidelines.md
[guidelines-tests]: guidelines.md#tests
```

File: docs/development/guidelines.md (hunk 8 above)
Comment (lars-reimann) [reply docs/development/guidelines.md:387]: It's fine as it's now.

**Review by lars-reimann — APPROVED** [review]
Excellent work, thanks!

[LGTM](https://lgtm.party/p/SJPH3OH/)
